httpbakery/agent/agent.go

```diff
@@ -106,121 +123,116 @@ func LoginCookie(req *http.Request) (username string, key *bakery.PublicKey, err
+type agent struct {
+	pathParts []string
+	username  string
+	key       *bakery.KeyPair
 }
 
-// SetCookie creates a cookie in jar which is suitable for performing agent
-// logins to u.
-//
-// If using SetUpAuth, it should not be necessary to use
-// this function.
-func SetCookie(jar http.CookieJar, u *url.URL, username string, pk *bakery.PublicKey) {
-	al := agentLogin{
-		Username:  username,
-		PublicKey: pk,
-	}
-	b, err := json.Marshal(al)
-	if err != nil {
-		// This shouldn't happen as the agentLogin type has to be marshalable.
-		panic(errgo.Notef(err, "cannot marshal cookie"))
+// Visitor is a httpbakery.Visitor that performs interaction using the
+// agent login protocol.
+type Visitor struct {
+	agents map[string][]agent
+}
+
+// AddAgent adds an agent to the visitor. The agent will have the given
+// username and key and will be used for requests that match the given
+// URL. If more than one agent matches a target URL then the most
+// specific one will be used. If an agent is added with an identical URL
+// as an existing agent then the existing agent will be replaced.
+func (v *Visitor) AddAgent(u *url.URL, username string, key *bakery.KeyPair) {
+	logger.Tracef("adding agent %s for %v", username, u)
+	if v.agents == nil {
+		v.agents = make(map[string][]agent)
 	}
-	v := base64.StdEncoding.EncodeToString(b)
-	jar.SetCookies(u, []*http.Cookie{{
-		Name:  cookieName,
-		Value: v,
-	}})
+	v.agents[u.Host] = insertAgent(v.agents[u.Host], agent{
+		pathParts: pathParts(u.Path),
+		username:  username,
+		key:       key,
+	})
 }
 
-// VisitWebPage creates a function that can be used with
-// httpbakery.Client.VisitWebPage. The function uses c to access the
-// visit URL. If no agent-login cookie has been configured for u an error
-// with the cause of ErrNoAgentLoginCookie will be returned. If the login
-// fails the returned error will be of type *httpbakery.Error. If the
-// response from the visitURL cannot be interpreted the error will be of
-// type *UnexpectedResponseError.
-//
-// If using SetUpAuth, it should not be necessary to use
-// this function.
-func VisitWebPage(c *httpbakery.Client) func(u *url.URL) error {
-	return func(u *url.URL) error {
-		err := ErrNoAgentLoginCookie
-		for _, c := range c.Jar.Cookies(u) {
-			if c.Name == cookieName {
-				err = nil
-				break
-			}
+func insertAgent(agents []agent, ag agent) []agent {
+	for i, a := range agents {
+		switch pathCmp(ag.pathParts, a.pathParts) {
+		case -1:
+			agents = append(agents, agent{})
+			copy(agents[i+1:], agents[i:])
+			agents[i] = ag
+			return agents
+		case 0:
+			agents[i] = ag
+			return agents
 		}
-		if err != nil {
-			return errgo.WithCausef(err, http.ErrNoCookie, "cannot perform agent login")
+	}
+	return append(agents, ag)
+}
+
+// pathParts splits a path up into an array of parts.
+func pathParts(path string) []string {
+	return strings.FieldsFunc(path, func(c rune) bool {
```

Comment:
Only in as much as they are part of the matching algorithm, so a can be sent to be if a is a prefix of b and a ends with a /, or the unmatched part of be starts with a /.